File: java/src/main/weekly/w299/CountNumberOfWaysToPlaceHouses.java
package weekly.w299;

/**
 * 2320. 统计放置房子的方式数：https://leetcode.cn/problems/count-number-of-ways-to-place-houses/
 */
public class CountNumberOfWaysToPlaceHouses {

    static final int MOD = (int) (1e9 + 7);

    public static int countHousePlacements(int n) {
        int prev = 1, next = 1;
        for (int i = 0; i < n; i++) {
            int oldNext = next;
            next = (prev + next) % MOD;
            prev = oldNext;
        }
        return (int) ((next * (long)next) % MOD);
    }

    public static void main(String[] args) {
        System.out.println(countHousePlacements(1)); // 4
        System.out.println(countHousePlacements(2)); // 9
    }

}

File: c/count-number-of-ways-to-place-houses.c
/**
 * 2320. 统计放置房子的方式数：https://leetcode.cn/problems/count-number-of-ways-to-place-houses
 */

#define MOD 1000000007

int countHousePlacements(int n) {
    int f[n + 1][2];
    f[0][0] = 1;
    f[0][1] = 0;
    for (int i = 1; i <= n; i++) {
        f[i][0] = (f[i - 1][0] + f[i - 1][1]) % MOD;
        f[i][1] = f[i - 1][0];
    }
    long long ans = (f[n][0] + f[n][1]) % MOD;
    ans = ans * ans % MOD;
    return ans;
}

int countHousePlacements2(int n) {
    long long f0 = 1, f1 = 0;
    for (int i = 1; i <= n; i++) {
        long long newF = (f0 + f1) % MOD;
        f1 = f0;
        f0 = newF;
    }
    long long ans = (f0 + f1) % MOD;
    ans = ans * ans % MOD;
    return ans;
}